### ISSUE TITLE
The span for “trait bound not satisfied” is weird

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#41817]
For following code

```
pub trait Arbitrary: Sized + 'static {}

impl<'a, A> Arbitrary for ::std::borrow::Cow<'a, A> {}
```

The error is

```
error[E0277]: the trait bound `A: std::clone::Clone` is not satisfied
 --> <anon>:3:13
  |
3 | impl<'a, A> Arbitrary for ::std::borrow::Cow<'a, A> {}
  |             ^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `A`
  |
  = help: consider adding a `where A: std::clone::Clone` bound
  = note: required because of the requirements on the impl of `std::borrow::ToOwned` for `A`
  = note: required because it appears within the type `std::borrow::Cow<'a, A>`
  = note: required by `Arbitrary`
```

Note the span is pointing at `Arbitrary`, despite error really being about the generic bound `A`. It should point at `A` instead.

**Comment by nox** [comment]
It is kinda a papercut in the case of trait bounds on associated types.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b106643284f7eb37ce6a6f087743952d

```rust
struct NotSync(*mut ());

trait AssociatedSync {
    type ShouldBeSync: Sync;
}

struct Dummy;

impl AssociatedSync /* span error points here */ for Dummy {
    type ShouldBeSync = NotSync /* whereas it should probably point here */;
}
```

**Comment by LukasKalbertodt** [comment]
I just attempted to fix it, but I already spent almost two hours into finding the relevant spot in the codebase and gave up for now. Can anyone give me some hints which functions might be relevant? I already roughly understand the functions involved in error reporting and am pretty sure that the wrong span ends up in an `ObligationCause`. But I have no idea where this span is created. In other words: where the obligation is checked.
